core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java

```diff
@@ -155,9 +156,20 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
       writer.commitAndClose();
     }
 
-    partitionLengths =
-      writePartitionedFile(shuffleBlockResolver.getDataFile(shuffleId, mapId));
-    shuffleBlockResolver.writeIndexFile(shuffleId, mapId, partitionLengths);
+    File output = shuffleBlockResolver.getDataFile(shuffleId, mapId);
+    final File tmp = new File(output.getAbsolutePath() + "." + UUID.randomUUID());
+    partitionLengths = writePartitionedFile(tmp);
+    if (!output.exists()) {
```

Comment:
I dont' think you can do this and still support SPARK-4085 -- regenerating the output if one of the shuffle files goes completely missing.  Because if the index file goes missing, and the data file is still there, with this logic you'll always never regenerate the shuffle output.  But maybe SPARK-4085 is not worth it ...